File: LR0/src/Types/Users/FractionUserType.java
package Types.Users;

import Comparator.Comparator;
import Comparator.FractionComparator;
import Types.FractionType;

import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Random;

public class FractionUserType implements UserType {

    private static final int MIN_INTEGER_PART = 0;
    private static final int MAX_INTEGER_PART = 10;
    private static final int MIN_NUMERATOR = 0;
    private static final int MAX_NUMERATOR = 1000;
    private static final int MIN_DENOMINATOR = 1;
    private static final int MAX_DENOMINATOR = 1000;

    @Override
    public String typeName() {
        return "Fraction";
    }

    @Override
    public Object create() throws InstantiationException {
        Random random = new Random();
        Boolean isNegative = random.nextBoolean();
        Integer partInteger = random.nextInt(MAX_INTEGER_PART - MIN_INTEGER_PART) + MIN_INTEGER_PART;
        Integer numerator = random.nextInt(MAX_NUMERATOR - MIN_NUMERATOR) + MIN_NUMERATOR;
        Integer denominator = random.nextInt(MAX_DENOMINATOR - MIN_DENOMINATOR) + MIN_DENOMINATOR;
        return new FractionType(isNegative, partInteger, numerator, denominator);
    }

    @Override
    public Object clone(Object object) throws InstantiationException {
        return new FractionType(((FractionType) object).getIsNegative(), ((FractionType) object).getIntegerPartValue(), ((FractionType) object).getNumeratorValue(), ((FractionType) object).getDenominatorValue());
    }

    @Override
    public Object readValue(InputStreamReader in) {
        try {
            return in.read();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Object parseValue(String ss) throws InstantiationException {
        String[] parts = ss.split("/");
        Integer left = null;
        Integer mid = null;
        Integer right = null;
        boolean isNegative = false;

        try {
            String[] leftMid = parts[0].split("\\.");
            left = Integer.parseInt(leftMid[0]);
            mid = Integer.parseInt(leftMid[1]);
            right = Integer.parseInt(parts[1]);
        } catch (NumberFormatException | ArrayIndexOutOfBoundsException e) {
            System.out.println("Error: Unable to parse the input.");
        }

        if (left != null && mid != null && right != null) {
            if (left < 0) {
                isNegative = true;
                left = -left;
            }
            return new FractionType(isNegative, left, mid, right);
        } else {
            return null;
        }
    }

    @Override
    public Comparator getTypeComparator() {
        return new FractionComparator();
    }

    @Override
    public String toString(Object object) {
        return object.toString();
    }

}

File: LR1/src/main/scala/types/users/FractionUserType.scala
package types.users

import comparator.FractionComparator
import types.FractionType
import java.io.IOException
import java.io.InputStreamReader
import java.util.Random

object FractionUserType {
  private val MIN_INTEGER_PART = 0
  private val MAX_INTEGER_PART = 10
  private val MIN_NUMERATOR = 0
  private val MAX_NUMERATOR = 1000
  private val MIN_DENOMINATOR = 1
  private val MAX_DENOMINATOR = 1000
}

class FractionUserType extends UserType {
  override def typeName = "Fraction"

  @throws[InstantiationException]
  override def create: AnyRef = {
    val random = new Random
    val isNegative = random.nextBoolean
    val partInteger = random.nextInt(FractionUserType.MAX_INTEGER_PART - FractionUserType.MIN_INTEGER_PART) + FractionUserType.MIN_INTEGER_PART
    val numerator = random.nextInt(FractionUserType.MAX_NUMERATOR - FractionUserType.MIN_NUMERATOR) + FractionUserType.MIN_NUMERATOR
    val denominator = random.nextInt(FractionUserType.MAX_DENOMINATOR - FractionUserType.MIN_DENOMINATOR) + FractionUserType.MIN_DENOMINATOR
    new FractionType(isNegative, partInteger, numerator, denominator)
  }

  @throws[InstantiationException]
  override def clone(`object`: AnyRef) = new FractionType(`object`.asInstanceOf[FractionType].getIsNegative, `object`.asInstanceOf[FractionType].getIntegerPartValue, `object`.asInstanceOf[FractionType].getNumeratorValue, `object`.asInstanceOf[FractionType].getDenominatorValue)

  override def readValue(in: InputStreamReader): Any = try in.read
  catch {
    case e: IOException =>
      throw new RuntimeException(e)
  }

  @throws[InstantiationException]
  override def parseValue(ss: String): AnyRef = {
    val parts = ss.split("/")
    var left: Integer = null
    var mid: Integer = null
    var right: Integer = null
    var isNegative = false
    try {
      val leftMid = parts(0).split("\\.")
      left = leftMid(0).toInt
      mid = leftMid(1).toInt
      right = parts(1).toInt
    } catch {
      case _: NumberFormatException | _: ArrayIndexOutOfBoundsException =>
        System.out.println("Error: Unable to parse the input.")
    }
    if (left != null && mid != null && right != null) {
      if (left < 0) {
        isNegative = true
        left = -left
      }
      new FractionType(isNegative, left, mid, right)
    }
    else null
  }

  override def getTypeComparator = new FractionComparator

  override def toString(`object`: AnyRef): String = `object`.toString
}